source/game_sa/CTxdStore.cpp

```diff
@@ -6,127 +6,325 @@ Do not delete this comment block. Respect others' work!
+    if (!txd)
+        return;
+
+    if (txd->m_pRwDictionary)
+        RwTexDictionaryDestroy(txd->m_pRwDictionary);
+    ms_pTxdPool->Delete(txd);
+}
 
 // remove txd
 // 0x731E90
-void CTxdStore::RemoveTxd(int index) { ((void(__cdecl*)(int))0x731E90)(index); }
+void CTxdStore::RemoveTxd(int32_t index) {
+    TxdDef* txd = ms_pTxdPool->GetAt(index);
+    if (!txd)
+        return;
+
+    if (txd->m_pRwDictionary) {
+        RwTexDictionaryForAllTextures(txd->m_pRwDictionary, RemoveIfRefCountIsGreaterThanOne, nullptr);
+        RwTexDictionaryDestroy(txd->m_pRwDictionary);
+    }
+    if (txd->m_wParentIndex != -1)
+        RemoveRef(txd->m_wParentIndex);
+
+    txd->m_pRwDictionary = nullptr;
+}
 
 // add reference for txd
 // 0x731A00
-TxdDef* CTxdStore::AddRef(int index) { return ((TxdDef * (__cdecl*)(int))0x731A00)(index); }
+void CTxdStore::AddRef(int32_t index) {
+    TxdDef* txd = ms_pTxdPool->GetAt(index);
+    if (txd)
+        txd->m_wRefsCount++;
+}
 
 // remove reference, if references count is 0, then delete txd
 // 0x731A30
-void CTxdStore::RemoveRef(int index) { ((void(__cdecl*)(int))0x731A30)(index); }
+void CTxdStore::RemoveRef(int32_t index) {
+    TxdDef* txd = ms_pTxdPool->GetAt(index);
+    if (!txd)
+        return;
+
+    if (--txd->m_wRefsCount <= 0)
+        CStreaming::RemoveModel(index + RESOURCE_ID_TXD);
+}
 
 // remove reference without deleting
 // 0x731A70
-TxdDef* CTxdStore::RemoveRefWithoutDelete(int index) { return ((TxdDef * (__cdecl*)(int))0x731A70)(index); }
+void CTxdStore::RemoveRefWithoutDelete(int32_t index) {
+    TxdDef* txd = ms_pTxdPool->GetAt(index);
+    if (txd)
+        txd->m_wRefsCount--;
+}
 
 // get references count
 // 0x731AA0
-unsigned int CTxdStore::GetNumRefs(int index) { return ((unsigned int(__cdecl*)(int))0x731AA0)(index); }
+int32_t CTxdStore::GetNumRefs(int32_t index) {
+    TxdDef* txd = ms_pTxdPool->GetAt(index);
+    return txd ? txd->m_wRefsCount : 0;
+}
+
+RwTexDictionary* CTxdStore::GetTxdParent(RwTexDictionary* txd) {
+    return *PLUGINOFFSET(RwTexDictionary*, txd, TexDictionaryLinkPluginOffset);
+}
+
+void CTxdStore::SetTxdParent(RwTexDictionary* txd, RwTexDictionary* parent) {
+    *PLUGINOFFSET(RwTexDictionary*, txd, TexDictionaryLinkPluginOffset) = parent;
+}
+
+// 0x731D50
+void CTxdStore::SetupTxdParent(int32_t index) {
+    TxdDef* txdcur = ms_pTxdPool->GetAt(index);
+    if (!txdcur || txdcur->m_wParentIndex == -1)
+        return;
+    TxdDef* txdpar = ms_pTxdPool->GetAt(txdcur->m_wParentIndex);
+    if (!txdpar)
+        return;
+
+    SetTxdParent(txdcur->m_pRwDictionary, txdpar->m_pRwDictionary);
+    AddRef(txdcur->m_wParentIndex);
+}
+
+// 0x731720
+RwTexture* CTxdStore::TxdStoreFindCB(const char* name) {
+    RwTexDictionary* txd = RwTexDictionaryGetCurrent();
+    for (; txd; txd = GetTxdParent(txd)) {
+        if (RwTexture* tex = RwTexDictionaryFindNamedTexture(txd, name))
+            return tex;
+    }
+    return nullptr;
+}
```

Comment:
```suggestion
RwTexture* CTxdStore::TxdStoreFindCB(const char* name) {
    RwTexDictionary* txd = RwTexDictionaryGetCurrent();
    for (; txd; txd = GetTxdParent(txd)) {
        RwTexture* tex = RwTexDictionaryFindNamedTexture(txd, name)
        if (tex)
            return tex;
    }
    return nullptr;
}
```

`if (RwTexture* tex = RwTexDictionaryFindNamedTexture(txd, name))` doesn't compile in C++03